### ISSUE TITLE
Parser context is tracked in different places when parsing top-level decls vs. parsing statements inside function bodies.

### ISSUE THREAD

**Issue by mayshukla** [ualberta-smr/varclang#2]
This is causing bugs where the parser is failing to join and also leads to lookups from the wrong contexts.

**Comment by mayshukla** [comment]
Parser is fails to join when for this test program:

```cpp
#ifdef A
int main() {

  #ifdef B
    return 0;
  #endif

}
#endif
```

**Comment by mayshukla** [comment]
Example of lookup from wrong context:
```cpp
int main() {

  #ifdef A
    int i = 0;
  #else
    int i = 1;
  #endif

  return i;
}
```

VarClang output:
```
sema_inside_function.cpp:5:9: error: redefinition of 'i'
    int i = 1;
        ^
sema_inside_function.cpp:3:9: note: previous definition is here
    int i = 0;
        ^
```

There should be no errors since the definitions are in mutually exclusive contexts.

**Comment by mayshukla** [comment]
Another example of lookup from wrong context:

```cpp
#ifdef A
  int a = 0;
#endif

int main() {
  #ifdef B
    return a;
  #else
    return 0;
  #endif
}
```

VarClang output:
```
10.cpp:7:10: error: a is not defined in context ~A
  return a;
```

TypeChef output:
```
#type checking
  - I [(def(B)&!def(A))] file 10.cpp:7:9--file 10.cpp:7:10
    a undeclared (only under condition def(A))
```

Note that TypeChef's error message is more specific.

**Comment by mayshukla** [comment]
Fixed by pull request #11 

Will need to add more information to diagnostics to fix the example where varclang differs from typechef. Created a new issue for this
